### DISCUSSION TITLE
Combining regular and biased meshes in 2D and 3D.

### DISCUSSION THREAD

**Discussion by EdSheltonUKAEA** [idaholab/moose#26216] [Q&A Meshing]
I need some help and advice on how to create efficient meshes in 2D and 3D, if possible using the built-in MOOSE meshing tools.  (But if there's an easy-to-learn alternative which I can use to create a mesh file which I then pull in to my MOOSE input file that could work too.)

I've really only worked in 1D with MOOSE before, almost exclusively using GeneratedMeshGenerator with combinations of regular and biased meshes.

But now I need 2D and 3D models initially for squares and cubes (or at least a quadrant of a square and an octant of a cube assuming symmetry at mid-point lines and planes).  GeneratedMeshGenerator explicitly says it can do this and I can produce a simple squares, rectangles or cubiod boxes.  But I want to be able to model concentric squares within squares and boxes within boxes for different mesh blocks and materials.  These would have finer or biased meshes close to the interfaces between layers.  I can't see a quick way to do this with GeneratedMeshGenerator as I would have to stitch together a fairly complicated arrangement of rectangles or cuboids.

And another probably more important challenge is that a 1D model with 1000 elements (and solves quite quickly) when extended to be the same in 2D and 3D gives 10^6 and 10^9 elements respectively.  This makes solving really slow (especially for 3D) and seems very inefficient as the elements could be much coarser further from the centre (similar to a cylindrical or spherical radial mesh).  So I think I need to try some alternative elements types, but again have no experience doing this having just used GeneratedMeshGenerator in 1D with default settings.

Any tips or suggestions on what mesh types and tools I could start using to do this?  Any examples of this type of mesh which others have built?  Is there a useful tutorial or video anywhere?  

Thanks in advance for any help!

**Comment by GiudGiud** [comment]
Hello

GeneratedMeshGenerator is indeed to simplistic if you want gradings in the cell density across the geometry.
However, CartesianMeshGenerator will let you have different densities. 
https://mooseframework.inl.gov/source/meshgenerators/CartesianMeshGenerator.html

It still is not the best, there will be regions that are over-discretized all the way to the boundary, but it will be better.

I think the best way to do these boxes with a center that is much finer than the outer rectangular shells is a combination of the CartesianMG with different subdomain IDs in regions and a uniform grid + a RefineBlockGenerator 
https://mooseframework.inl.gov/source/meshgenerators/RefineBlockGenerator.html

This will let you select the subdomain near the center and apply mesh refinement there. The mesh wont be conformal anymore, but it's in a way that is supported by MOOSE/libmesh (h-refinement)

Guillaume
